packages/tokenbank_v1/contracts/impl/BaseWallet.sol

```diff
@@ -0,0 +1,242 @@
+  limitations under the License.
+*/
+pragma solidity ^0.5.11;
+
+import "../iface/Wallet.sol";
+import "../iface/Module.sol";
+
+import "../lib/ERC20.sol";
+import "../lib/NamedAddressSet.sol";
+import "../lib/ReentrancyGuard.sol";
+
+
+/// @title BaseWallet
+/// @dev This contract provides basic implementation for a Wallet.
+///
+/// @author Daniel Wang - <daniel@loopring.org>
+///
+/// The design of this contract is inspired by Argent's contract codebase:
+/// https://github.com/argentlabs/argent-contracts
+contract BaseWallet is Wallet, NamedAddressSet, ReentrancyGuard
+{
+    string private constant MODULE = "__MODULE__";
+    string private constant ERC20_TRANSFER = "transfer(address,uint256)";
+
+    mapping (bytes4  => address) internal methodToModule;
+
+    /// @dev Emitted when the wallet received Ether.
+    event Received(
+        address indexed sender,
+        uint    value,
+        bytes   data
+    );
+
+    modifier onlyOwner
+    {
+        require(msg.sender == owner, "NOT_A_OWNER");
+        _;
+    }
+
+    modifier onlyModule
+    {
+        require(isAddressInSet(MODULE, msg.sender), "NOT_A_MODULE");
+        _;
+    }
+
+    function init(
+        address            _owner,
+        address[] calldata _modules
+        )
+        external
+        nonReentrant
+    {
+        require(owner == address(0) && numAddressesInSet(MODULE) == 0, "INITIALIZED_ALREADY");
+        require(_owner != address(0), "ZERO_ADDRESS");
+        require(_modules.length > 0, "EMPTY_MODULES");
+
+        owner = _owner;
+
+        emit Initialized(owner);
+
+        for(uint i = 0; i < _modules.length; i++) {
+            address module = _modules[i];
+            require(module != address(0), "NULL_MODULE");
+            addAddressToSet(MODULE, module);
+            Module(module).init(address(this));
+            emit ModuleAdded(module);
+        }
+    }
+
+    function addModule(address _module)
+        external
+        onlyModule
+        nonReentrant
+    {
+        require(_module != address(0), "NULL_MODULE");
+        addAddressToSet(MODULE, _module);
+        Module(_module).init(address(this));
+        emit ModuleAdded(_module);
+    }
+
+    function removeModule(address _module)
+        external
+        onlyModule
+        nonReentrant
+    {
+        require(numAddressesInSet(MODULE) > 1, "PROHIBITED");
+        removeAddressFromSet(MODULE, _module);
+        emit ModuleRemoved(_module);
+    }
+
+    function getModules()
+        public
+        view
+        returns (address[] memory)
+    {
+        return addressesInSet(MODULE);
+    }
+
+    function hasModule(address _module)
+        public
+        view
+        returns (bool)
+    {
+        return isAddressInSet(MODULE, _module);
+    }
+
+    function bindStaticMethod(bytes4 _method, address _module)
+        external
+        onlyModule
+        nonReentrant
+    {
+        require(methodToModule[_method] != address(0), "BAD_MODULE");
```

Comment:
Maybe this needs to be `require(methodToModule[_method] == address(0)`? Or it may not be a bad idea to remove the check all together so that this method can also be used to unbind methods by setting `_module == address(0)`?

Reply:
As you suggested:
```solidity
require(_method != bytes4(0), "BAID_METHOD");
        require(
            _module == address(0) || methodToModule[_method] == address(0),
            "BAD_MODULE"
);

```
this will allow unbinding of methods.